hubbubs/abstract.py

```diff
@@ -0,0 +1,170 @@
+from django.db import models
+from django.contrib.sites.models import Site
+from django.utils.translation import ugettext_lazy as _
+from django.utils import crypto, timezone
+from datetime import timedelta
+
+from requests import (
+    codes, exceptions as req_exceptions, request as request_sender)
+from urlparse import urljoin, urlparse
+
+from .exceptions import SubscriptionError
+from .settings import USE_SSL, LEASE_SECONDS
+
+
+class AbstractSubscription(models.Model):
+
+    topic = models.URLField(_('Topic'), max_length=255)
+    hub = models.URLField(_('Hub'), max_length=255)
+    # this token should be changed only when a new subscription is issued;
+    #   the token is also used by the hub when it reaches lease_expiration
+    #   and sends a new verification request
+    verify_token = models.CharField(
+        _('Verify Token'), max_length=255, null=True, blank=True)
+    lease_expiration = models.DateTimeField(
+        _('Lease expiration'), editable=False, null=True, blank=True)
+    secret = models.CharField(
+        _('Secret'), max_length=255, null=True, blank=True)
+
+    # active and inactive statuses are set only when verification succeeded
+    INACTIVE = 0
+    ACTIVE = 1
+    # verify status is set only when a async
+    #   subscribe/unsubscribe action was submitted
+    VERIFYING = 2
+    # rejection status is set when async verification is rejected
+    SUB_REJECTED = 3
+    UNSUB_REJECTED = 4
+    STATUS_CODES = {
+        INACTIVE: u'inactive',
+        ACTIVE: u'active',
+        VERIFYING: u'verifying',
+        SUB_REJECTED: u'subscribe action rejected',
+        UNSUB_REJECTED: u'unsubscribe action rejected',
+    }
+    status = models.PositiveSmallIntegerField(
+        editable=False, choices=STATUS_CODES.items(), default=INACTIVE)
+    # used for debugging purposes and will show when the last activity
+    #   for this subscription happened
+    updated_at = models.DateTimeField(auto_now=True)
+    site = models.ForeignKey(Site, blank=True, null=True)
+
+    class Meta:
+        abstract = True
+        unique_together = (("topic", "site"), )
+
+    def uses_ssl(self):
+        return USE_SSL
+
+    @property
+    def callback_url(self):
+        raise NotImplementedError()
+
+    def full_callback_url(self):
+        domain = (self.site or Site.objects.get_current()).domain
+        return urljoin("http://" + domain, self.callback_url)
+
+    def subscribe(self, verify_mode=u'sync', lease_seconds=None):
+        return self._send_subscription(
+            u'subscribe',
+            verify_mode=verify_mode,
+            lease_seconds=lease_seconds
+        )
+    subscribe.alters_data = True
```

Comment:
Django is using it when rendering templates: {{obj.subscribe}}. The template system won’t call a variable if it has alters_data=True set. https://docs.djangoproject.com/en/1.4/ref/templates/api/#rendering-a-context